cpp/common/src/codingstandards/cpp/alertreporting/DeduplicateMacroResults.qll

```diff
@@ -0,0 +1,379 @@
+ *   }
+ *
+ *   module DeduplicateFooInMacrosConfig implements DeduplicateMacroConfigSig<InvalidFoo> {
+ *     string describe(InvalidFoo foo) { result = ... }
+ *   }
+ *
+ *   import DeduplicateMacroResults<InvalidFoo, DeduplicateFooInMacrosConfig> as DeduplicateFooInMacros;
+ * ```
+ *
+ * This module exposes the following classes:
+ *  - `PrimaryMacroSameResultElementInAllInvocations extends Macro`: Every invocation of this macro
+ *    generates an `InvalidFoo` which stringifies to the same thing. Use the method
+ *    `getResultElementDescription()` to get that shared string.
+ *  - `PrimaryMacroDifferentResultElementInAllInvocations extends Macro`: Every invocation of this
+ *    macro generates an `InvalidFoo`, but they do not all stringify to the same thing. Use the
+ *    method `getExampleResultElement()` to get an *single* example `InvalidFoo` to help users fix
+ *    and understand the issue.
+ *  - `IsolatedMacroExpansionWithResultElement extends MacroInvocation`: An invocation of a macro
+ *    which in this particular instance generates an `InvalidFoo`, while other invocations of the
+ *    same macro do not.
+ *
+ * The three above classes all attempt to resolve to the most *specific* macro to the issue at
+ * hand. For instance, if macro `M1` calls macro `M2` which expands to an `InvalidFoo`, then the
+ * problem may be with `M2` (it is the most specific macro call here), or the problem may be with
+ * `M2` (if all expansions of `M2` generate an `InvalidFoo` but not all expansions of `M1` do so).
+ *
+ * ## Generating Report Objects
+ *
+ * This module also can be used to more easily report issues across these cases, by implementing
+ * `MacroReportConfigSig` and importing `DeduplicateMacroResults::Report::ReportResultInMacro`.
+ *
+ * ```
+ *   module InvalidFooInMacroReportConfig implements MacroReportConfigSig<InvalidFoo> {
+ *
+ *     // ***Take care that usage of $@ is correct in the following predicates***!!!!
+ *     bindingset[description]
+ *     string getMessageSameResultInAllExpansions(Macro m, string description) {
+ *       result = "Macro " + m.getName() + " always has invalid foo " + description
+ *     }
+ *
+ *     string getMessageVariedResultInAllExpansions(Macro m) {
+ *       result = "Macro " + m.getName() + " always has invalid foo, for example '$@'."
+ *     }
+ *
+ *     string getMessageResultInIsolatedExpansion(InvalidFoo foo) {
+ *       result = "Invocation of macro $@ has invalid foo '" + foo.getName() + "'."
+ *     }
+ *   }
+ *
+ *   import DeduplicateFooInMacros::Report<InvalidFooInMacroReportConfig> as Report;
+ *
+ *   from Report::ReportResultInMacro report
+ *     where not excluded(report.getPrimaryElement(), ...)
+ *   select report.getPrimaryElement(), report.getMessage(), report.getOptionalPlaceholderLocation(),
+ *     report.getOptionalPlaceholderMessage()
+ * ```
+ *
+ * Note that this does *not* (currently) generate a result for elements not contained by a macro.
```

Comment:
I'm wondering if it would create a cleaner interface to do both in this library? That would reduce the number of queries required in this case to two queries, instead of four. It would also mean we could replace/integrate `MacroUnwrapper`, which is only currently used by one query.

Reply:
What was making this more complicated was `MacroExpansionWithOnlyUnusedObjectDefinition` which made it so some macros were being deduped and others weren't. I don't think that class was adding much UX value...probably some, but probably very marginal. I removed that class in favor of deduping all macro invocations the same way, which still has a case for when only particular invocations of macros result in unused vars. That allowed the deletion of all of this bloat, much better now!